sdk/core/azure-core/azure/core/rest/_rest_py3.py

```diff
@@ -262,16 +262,22 @@ def encoding(self) -> Optional[str]:
     def encoding(self, value: str) -> None:
         """Sets the response encoding"""
         self._encoding = value
+        self._text = None  # clear text cache
 
-    @property
-    def text(self) -> str:
-        """Returns the response body as a string"""
-        if self._text is None:
+    def text(self, encoding: Optional[str] = None) -> str:
+        """Returns the response body as a string
+
+        :param optional[str] encoding: The encoding you want to decode the text with. Can
+         also be set independently through our encoding property
+        :return: The response's content decoded as a string.
+        """
+        if self._text is None or encoding:
             content = self.content
             if not content:
                 self._text = ""
             else:
-                self._text = decode_to_text(self.encoding, self.content)
+                encoding_to_pass = encoding or self.encoding
+                self._text = decode_to_text(encoding_to_pass, self.content)
```

Comment:
Why use `self.content` when you fished out the value on line 275. Alternatively, why are you fishing out content on line 275? Do we want to save the instructions needed to access the attribute from the instance?

Reply:
Is it because accessing the content property has side-effects?

Reply:
Acccessign the `content` property will raise the `ResponseNotRead` errors if users mishandle. You ahve a good point, I don't have to store it in another variable, so I'll just stick to using `self.content`

Reply:
I agree with pylint that the side effects of accessing `self.content` are too subtle. You are going to access self.content anyways, why do you need it *also* call it here?

Reply:
sorry i'm not quite understanding your comment. Does the current code there look better to you?

Reply:
Why not just say:

```python
if not self.content:  # <- this will raise if data is not read. Line 275 is not needed. 
    self._text = ""
else:
    ...
```

Reply:
ah i think i was confused bc we're looking at diff code versions. What I currently have is
```python
if self._text is None or encoding:
    encoding_to_pass = encoding or self.encoding
    self._text = decode_to_text(encoding_to_pass, self.content)
return self._text
```
which i think solves the issue you brought up (i moved the empty content detection to `decode_to_text`